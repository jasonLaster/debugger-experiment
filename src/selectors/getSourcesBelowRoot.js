// @flow

import { getProjectDirectoryRoot, getSources } from "../selectors";
import type { OuterState } from "../reducers/types";
import type { Source } from "../types";
import { dropScheme } from "../utils/source";

export type RelativeSource = Source & {
  +relativeUrl: string
};

function getRelativeUrl(url, root) {
  const sourcePathSplit = url.split("/");

  // Rebuild the sourcePath
  let newSourcePath = sourcePathSplit.join("/");

  if (!root) {
    return dropScheme(url);
  }

  return url.slice(url.indexOf(root) + root.length);
}

function formatSource(source: Source, root): RelativeSource {
  return {
    ...source,
    relativeUrl: getRelativeUrl(source.url, root)
  };
}

/*
 */
export function getSourcesBelowRoot(state: OuterState) {
  const sources = getSources(state);
  const root = getProjectDirectoryRoot(state);

  return sources
    .valueSeq()
    .toJS()
    .filter(({ url }) => url && url.includes(root))
    .map(source => formatSource(source, root));
}
