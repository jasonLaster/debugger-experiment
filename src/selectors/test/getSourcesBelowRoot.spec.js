/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { getSourcesBelowRoot } from "../getSourcesBelowRoot";

import {
  createStore,
  selectors,
  actions,
  makeSource
} from "../../utils/test-head";

describe("getSourcesBelowRoot", () => {
  it("simple", async () => {
    const store = createStore({});
    const { dispatch, getState } = store;
    await dispatch(actions.newSource(makeSource("js/scopes.js")));
    await dispatch(actions.newSource(makeSource("lib/vendor.js")));
    dispatch(actions.setProjectDirectoryRoot("/js"));
    const filteredSources = getSourcesBelowRoot(getState());
    const firstSource = filteredSources[0];

    expect(firstSource.url).toEqual(
      "http://localhost:8000/examples/js/scopes.js"
    );
    expect(firstSource.relativeUrl).toEqual("/scopes.js");
  });
});
