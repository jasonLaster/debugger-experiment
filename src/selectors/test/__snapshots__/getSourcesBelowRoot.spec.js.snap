// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSourcesBelowRoot selector simple 1`] = `
Array [
  Object {
    "contentType": "",
    "error": undefined,
    "id": "js/scopes.js",
    "isBlackBoxed": false,
    "isPrettyPrinted": false,
    "isWasm": false,
    "loadedState": "unloaded",
    "relativeUrl": "/scopes.js",
    "sourceMapURL": undefined,
    "text": undefined,
    "url": "http://localhost:8000/examples/js/scopes.js",
  },
]
`;
